clash-prelude/tests/Clash/Tests/AsyncFIFOSynchronizer.hs

```diff
@@ -963,6 +970,149 @@ test6W7 = testW sync6W7
+                   )
+     , Signal wdom ( Bool
+                     -- Test fully done
+                   , Maybe Int
+                    -- Element written to FIFO
+                   )
+     )
+fifoOperations racts wacts = (bundle (rAllDone, rel), bundle (wAllDone, wel))
+ where
+  (rdata, rempty, wfull) =
+    asyncFIFOSynchronizer d3 wclk rclk noWRst noRRst (toEnable wen)
+      (toEnable ren) rinc wDataM
+  rclk = clockGen @rdom
+  wclk = clockGen @wdom
+  noRRst = unsafeFromHighPolarity @rdom (pure False)
+  noWRst = unsafeFromHighPolarity @wdom (pure False)
+  (wdone, wact) =
+    unbundle $ fromList $ P.zip (P.repeat False) wacts <> P.repeat (True, WNoOp)
+  (rdone, ract) =
+    unbundle $ fromList $ P.zip (P.repeat False) racts <> P.repeat (True, RNoOp)
+  (wen, wDataM) = unbundle $
+    liftA2 (\wact0 d -> case wact0 of
+                          Write      -> (True , Just d )
+                          WNoOp      -> (True , Nothing)
+                          WGated     -> (False, Nothing)
+                          GatedWrite -> (False, Just d )
+           ) wact wdata
+  wdata = regEn wclk noWRst enableGen 1 (isJust <$> wDataM) (wdata + 1)
+  (ren, rinc) = unbundle $
+    fmap (\ract0 -> case ract0 of
+                      Read       -> (True , True )
+                      RNoOp      -> (True , False)
+                      RGated     -> (False, False)
+                      GatedRead  -> (False, True)
+         ) ract1
+  mainDone = rdone .&&. unsafeSynchronizer wclk rclk wdone
+  -- Empty FIFO after main test
+  ract1 = mux mainDone (pure Read) ract
+  flushCnt = regEn rclk noRRst enableGen (1 :: Int) mainDone (flushCnt + 1)
+  -- Surely in 20 cycles we have seen all of the FIFO even if something has gone
+  -- wrong with the pointers
+  rAllDone = fmap (> 20) flushCnt
+  wAllDone = unsafeSynchronizer rclk wclk rAllDone
+  rel = mux (liftA3 (\en inc em -> en && inc && not em) ren rinc rempty)
+            (Just <$> rdata) (pure Nothing)
+  wel = mux (liftA3 (\en wdm fu -> en && isJust wdm && not fu) wen wDataM wfull)
+            (Just <$> wdata) (pure Nothing)
+
+fifoFunctionalTest
+  :: forall wdom rdom m
+   . Monad m
+  => DomProxy wdom
+  -> DomProxy rdom
+  -> NamedTest (PropertyT m ())
+fifoFunctionalTest wdom@DomProxy rdom@DomProxy =
+  NamedTest { namedTest=test0, testName = name }
+ where
+  name = testNameShowsPrec 0 (shows 'fifoFunctionalTest) wdom rdom ""
+  test0 = do
+    racts <- fmap P.concat . forAll . Gen.list (Range.linear 1 10) . Gen.list (Range.linear 0 20) $ Gen.enumBounded
+    wacts <- fmap P.concat . forAll . Gen.list (Range.linear 1 10) . Gen.list (Range.linear 0 20) $ Gen.enumBounded
+    let rels = catMaybes . P.map snd . P.takeWhile (not . fst) . sample . fst $
+                 fifoOperations @wdom @rdom racts wacts
+        wels = catMaybes . P.map snd . P.takeWhile (not . fst) . sample . snd $
+                 fifoOperations @wdom @rdom racts wacts
```

Comment:
This simulates `fifoOpperations` twice. I doubt that is what we want, given that the arguments to it are exactly the same. I'd split it up, something like:

```haskell
let
  results = fifoOperations @wdom @rdom racts wacts
  catValids = catMaybes . P.map snd . P.takeWhile (not . fst)
  rels = catValids (fst (sample results))
  wels = catValids (snd (sample results))
```

Reply:
Ah yes, of course, good point.